// @jsmini/is/dist/index.aio.js
/*!
 * is 0.7.1 (https://github.com/jsmini/is)
 * API https://github.com/jsmini/is/blob/master/doc/api.md
 * Copyright 2017-2018 yanhaijing. All Rights Reserved
 * Licensed under MIT (https://github.com/jsmini/is/blob/master/LICENSE)
 */

(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
	typeof define === 'function' && define.amd ? define(['exports'], factory) :
	(factory((global.jsmini_is = {})));
}(this, (function (exports) { 'use strict';

/*!
 * type 0.4.1 (https://github.com/jsmini/type)
 * API https://github.com/jsmini/type/blob/master/doc/api.md
 * Copyright 2017-2018 jsmini. All Rights Reserved
 * Licensed under MIT (https://github.com/jsmini/type/blob/master/LICENSE)
 */

const toString = Object.prototype.toString;

function type(x) {
    if(x === null){
        return 'null';
    }

    const t= typeof x;

    if(t !== 'object'){
        return t;
    }

    let c;
    try {
        c = toString.call(x).slice(8, -1).toLowerCase();
    } catch(e) {
        return 'object';
    }

    if(c !== 'object'){
        return c;
    }

    if(x.constructor == Object){
        return c;
    }

    try {
        // Object.create(null)
        if (Object.getPrototypeOf(x) === null || x.__proto__ === null) {
            return 'object';
        }

        return 'unknown';
    } catch(e) {
        // ie下无Object.getPrototypeOf
        return 'unknown';
    }
}

function isInRange(x, min, max) {
    x = +x;
    min = +min;
    max = +max;

    // x 不是数字，则返回false
    if (isNaN(x)) return false;

    // min 或 max 不传，则认为不设置下限或上限
    return (!isNaN(min) ? x >= min : true) && (!isNaN(max) ? x <= max : true);
}

function isNumber(x, min, max) {
    return type(x) === 'number' && isInRange(x, min, max);
}

function isInteger(x, min, max) {
    return parseInt(x, 10) === x  && isInRange(x, min, max);
}

function isInt(x) {
    // -2^31 ~ 2^31-1
    return isInteger(x, -2147483648, 2147483647);
}

function isBoolean(x) {
    return type(x) === 'boolean';
}

function isString(x) {
    return type(x) === 'string';
}

function isEmptyString(x) {
    if (!isString(x)) return false;

    return x.replace(/(^\s*|\s*$)/g, '') === '';
}

function isNull(x) {
    return type(x) === 'null';
}

function isUndefined(x) {
    return type(x) === 'undefined';
}

function isObject(x) {
    return type(x) === 'object';
}

function isFunction(x) {
    return type(x) === 'function';
}

const isArray = isFunction(Array.isArray) ? Array.isArray : function isArray(x) {
    return type(x) === 'array';
};

exports.isInRange = isInRange;
exports.isNumber = isNumber;
exports.isInteger = isInteger;
exports.isInt = isInt;
exports.isBoolean = isBoolean;
exports.isString = isString;
exports.isEmptyString = isEmptyString;
exports.isNull = isNull;
exports.isUndefined = isUndefined;
exports.isObject = isObject;
exports.isFunction = isFunction;
exports.isArray = isArray;

})));
